File: 백준/Gold/1953. 팀배분/팀배분.py
import sys
from collections import deque
input = sys.stdin.readline


n=int(input().rstrip())
graph=[[]for _ in range(n+1)]
visit=[0]*(n+1)
color=[-1]*(n+1)
for l in range(1,n+1):
    e=list(map(int,input().split()))
    for i in range(1,len(e)):
        graph[l].append(e[i])

def bfs(l):
    global color
    q = deque()
    visit[l]=1
    q.append((l,0))
    color[l]=0

    while q:
        vertex,co=q.popleft()
        for t1 in graph[vertex]:
            if visit[t1]==0:
                next=-1
                if co==0:
                    next=1
                else:
                    next=0
                color[t1]=next
                visit[t1]=1
                q.append((t1,next))


for l in range(1,n+1):
    if visit[l]==0:
        bfs(l)


blue=[]
red=[]
for l in range(1,n+1):
    if color[l]==0:
        blue.append(l)
    else:
        red.append(l)

blue.sort()
red.sort()
print(len(blue))
print(*blue)
print(len(red))
print(*red)

File: 백준/Gold/1953. 팀배분/팀배분.java
import java.awt.*;
import java.beans.VetoableChangeListener;
import java.io.*;
import java.security.cert.CertificateParsingException;
import java.util.*;
import java.util.List;


public class Main {


    public static class go {
        int index, color;

        public go(int index, int color) {
            this.index = index;
            this.color = color;
        }
    }

    static ArrayList<Integer>[] list;

    static void bfs(int[] visit, int vertex) {
        Deque<go> q = new ArrayDeque<>();
        visit[vertex] = 1;
        q.add(new go(vertex, 1));
        while (!q.isEmpty()) {
            go a = q.poll();
            for (int v : list[a.index]) {
                if (visit[v] == 0) {
                    int next = 0;
                    if (a.color == 1)
                        next = 2;
                    else
                        next = 1;
                    q.add(new go(v, next));
                    visit[v]=next;
                }
            }
        }
    }


    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer s = new StringTokenizer(br.readLine(), " ");
        //청팀과 백팀으로 팀을 나누어 팀전을 진행하자
        int n = Integer.parseInt(s.nextToken());
        //사람들이 각각 싫어하는 사람들의 정보가 주어질시에 , 서로 싫어하는 사람은 같은 팀에 넣지 않으려 한다
        list = new ArrayList[n + 1];
        for (int i = 1; i < n + 1; i++) {
            list[i] = new ArrayList<>();
        }
        for (int i = 1; i <= n; i++) {
            s = new StringTokenizer(br.readLine(), " ");
            int leng = Integer.parseInt(s.nextToken());
            for (int j = 0; j < leng; j++) {
                list[i].add(Integer.parseInt(s.nextToken()));
            }
        }

        //이제 시작하자
        int visit[] = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            if (visit[i] == 0) {
                bfs(visit, i);
            }
        }
        List<Integer>blue=new ArrayList<>();
        List<Integer>white=new ArrayList<>();
        for(int i=1;i<=n;i++){
            if (visit[i]==1)
                blue.add(i);
            else
                white.add(i);
        }
        System.out.println(blue.size());
        for (int a:blue)
            System.out.print(a+" ");
        System.out.println();
        System.out.println(white.size());
        for (int a:white)
            System.out.print(a+" ");

    }


}